### PR TITLE
Exposing more methods in IMenuModel to facilitate context menu custom…

### PR DESCRIPTION
**Pull request by zkovacs1**
…izations

### CODE REVIEW

**Comment by amaitland** [comment]
Welcome! Few things before this can be merged, I'll comment inline.

File: CefSharp/IMenuModel.cs [hunk 1]
```diff
@@ -7,5 +7,10 @@ namespace CefSharp
 	public interface IMenuModel
 	{
 		bool Clear();
-	}
+        int GetCount();
```
Comment (amaitland) [inline CefSharp/IMenuModel.cs]: We map things like `GetCount` to being a `Count` property with a `getter`.

File: CefSharp/IMenuModel.cs [hunk 2]
```diff
@@ -7,5 +7,10 @@ namespace CefSharp
...
+        string GetLabelAt(int index);
+        int GetCommandIdAt(int index);
+        bool Remove(int index);
+        bool AddItem(int command_id, string label);
```
Comment (amaitland) [inline CefSharp/IMenuModel.cs]: `command_id` should be `commandId`

File: CefSharp.WinForms.Example/Handlers/MenuHandler.cs [hunk 3]
```diff
@@ -10,16 +10,24 @@ void IContextMenuHandler.OnBeforeContextMenu(IWebBrowser browserControl, IBrowse
...
         {
+            if (commandId == 20222)
+                browser.Reload(true);
             return false;
         }
 
         void IContextMenuHandler.OnContextMenuDismissed(IWebBrowser browserControl, IBrowser browser, IFrame frame)
         {
-            
+            return;
```
Comment (amaitland) [inline CefSharp.WinForms.Example/Handlers/MenuHandler.cs]: Please remove unnessicary changes.

File: CefSharp.WinForms.Example/Handlers/MenuHandler.cs [hunk 4]
```diff
@@ -10,16 +10,24 @@ void IContextMenuHandler.OnBeforeContextMenu(IWebBrowser browserControl, IBrowse
         {
             //To disable the menu then call clear
             // model.Clear();
+
+            bool removed = model.Remove(132); // Remove "View Source" option
```
Comment (amaitland) [inline CefSharp.WinForms.Example/Handlers/MenuHandler.cs]: I'm not in favor of removing `View Source` from the demo, please revert.

File: CefSharp.WinForms.Example/Handlers/MenuHandler.cs [hunk 5]
```diff
@@ -10,16 +10,24 @@ void IContextMenuHandler.OnBeforeContextMenu(IWebBrowser browserControl, IBrowse
...
+            if (removed)
+            {
+                model.AddItem(20222, "Reload"); // 20222 - our custom command id
+            }
         }
 
         bool IContextMenuHandler.OnContextMenuCommand(IWebBrowser browserControl, IBrowser browser, IFrame frame, IContextMenuParams parameters, int commandId, CefEventFlags eventFlags)
         {
+            if (commandId == 20222)
```
Comment (amaitland) [inline CefSharp.WinForms.Example/Handlers/MenuHandler.cs]: Add braces to all blocks, even single line ones.

File: CefSharp.Core/Internals/CefMenuModelWrapper.h [hunk 6]
```diff
@@ -35,5 +35,29 @@ namespace CefSharp
         {
             return _menu->Clear();
         }
+
+		virtual int GetCount() {
```
Comment (amaitland) [inline CefSharp.Core/Internals/CefMenuModelWrapper.h]: General thing, spaces not tabs

File: CefSharp.Core/Internals/CefMenuModelWrapper.h [hunk 7]
```diff
@@ -35,5 +35,29 @@ namespace CefSharp
...
+			return _menu->GetCount();
+		}
+
+		virtual String^ GetLabelAt(int index) {
```
Comment (amaitland) [inline CefSharp.Core/Internals/CefMenuModelWrapper.h]: Braces on new lines.

File: CefSharp.Core/Internals/CefMenuModelWrapper.h [hunk 8]
```diff
@@ -35,5 +35,29 @@ namespace CefSharp
...
+
+		virtual bool Remove(int index) {
+			return _menu->Remove(index);
+		}
+
+		virtual bool AddSeparator() {
+			return _menu->AddSeparator();
+		}
+
+		virtual bool AddItem(int command_id, String^ label) {
```
Comment (amaitland) [inline CefSharp.Core/Internals/CefMenuModelWrapper.h]: Rename to `commandId`

File: CefSharp.WinForms.Example/Handlers/MenuHandler.cs (hunk 4 above)
Comment (amaitland) [inline CefSharp.WinForms.Example/Handlers/MenuHandler.cs]: I think we probably need to expose the `enum` for commands, so they're a little more user friendly.

Comment (zkovacs1) [inline CefSharp.WinForms.Example/Handlers/MenuHandler.cs]: Hey,

Thanks for the quick and very detailed comments/suggestions!

I did some search trying to find the enum before committing my changes, but no luck :/ I do not like inline values either.

Regards,
Zoltan

Comment (amaitland) [inline CefSharp.WinForms.Example/Handlers/MenuHandler.cs]: Will need to add a new `enum` based on http://magpcss.org/ceforum/apidocs3/projects/%28default%29/cef_menu_id_t.html

Comment (amaitland) [inline CefSharp.WinForms.Example/Handlers/MenuHandler.cs]: (That link is slightly outdated, will need to look in the `include` files for up to date values).

File: CefSharp/IMenuModel.cs [hunk 9]
```diff
@@ -4,8 +4,13 @@
 
 namespace CefSharp
 {
-	public interface IMenuModel
-	{
-		bool Clear();
-	}
+    public interface IMenuModel
+    {
+        int Count { get; }
```
Comment (amaitland) [inline CefSharp/IMenuModel.cs]: Do you mind adding `xml` documentation to these? Taking to improve documentation as a whole :+1:

**Comment by amaitland** [comment]
Should we expose the other `CefMenuModel` methods? Is the subset you have exposed enough for general usage?

File: CefSharp/IMenuModel.cs (hunk 9 above)
Comment (zkovacs1) [inline CefSharp/IMenuModel.cs]: Will do my best. Trapped in other stuff right now.

Regards,
Zoltan

**Comment by zkovacs1** [comment]
The exposed subset is enough to add/remove menu items.

For the sake of completeness however, on longer  term perhaps is better to expose the other methods, too. (Why CefSharp provide less functionality than Cef itself) ?

Regards,
Zoltan

**Comment by amaitland** [comment]
Merged with following changes

https://github.com/cefsharp/CefSharp/commit/fa7728b0d0377ea7dfcc4e5d0efb776775892469
https://github.com/cefsharp/CefSharp/commit/fc4307629b6cf9e7f81d8f08d8d1475ae46bf7ee

**Pull request merged**